course/api/views.py

```diff
@@ -195,20 +198,60 @@ class CourseExercisesViewSet(NestedViewSetMixin,
     lookup_url_kwarg = 'exercisemodule_id'
     lookup_value_regex = REGEX_INT
     parent_lookup_map = {'course_id': 'course_instance.id'}
-    serializer_class = CourseModuleSerializer
 
-    def get_queryset(self) -> QuerySet[CourseModule]:
-        return (
-            CourseModule.objects
-            .get_visible(self.request.user)
-            .prefetch_related(
-                Prefetch('learning_objects', BaseExercise.objects.all(), 'exercises')
-            )
-            .all()
-        )
+    # this converts a single module into a response dictionary
+    def __module_to_dict(self, module, **kwargs):
+        # if specific module then detail otherwise all modules (list)
+        api_url_string = "course-exercises-detail" if 'exercisemodule_id' in kwargs else "course-exercises-list"
+        api_url = build_aplus_url(api_reverse(api_url_string, kwargs=kwargs))
+        module_html_url = build_aplus_url(module['link'])
+        module_dictionary = {
+            'id': module['id'],
+            'url': api_url,
+            'html_url': module_html_url,
+            'display_name': module['name'],
+            'is_open': CourseModule.is_open_helper(module['reading_opening_time'], module['opening_time'],
+                                                   module['closing_time']
+                                                   ),
+            'exercises': [],
+        }
+        # add this module to cached exercises
+        for exercise in filter(lambda ex: ex['type'] == 'exercise', module['children']):
+            # extract all the relevant fields of the exercise
+            api_url = build_aplus_url(api_reverse(
+                'exercise-detail', kwargs={'exercise_id': exercise['id']}))
+
+            # check if there is a xx.yy pattern at the start of the name (e.g. 6.2 Hello Python). Also matches 1.
+            hierarchical_name = exercise['name'] if re.match(
+                "^[0-9]+.[0-9]*", exercise['name']) else f"{exercise['number']} {exercise['name']}"
```

Comment:
Note that course instance settings include the possibility for Roman numerals (I, II, III, IV, V, VI, ...).

This regex does not cover the Roman numerals yet.

https://github.com/apluslms/a-plus/blob/9ed509f62eaaf7d7b87c1238d54e04b8a9bd0ae4/exercise/exercise_models.py#L230-L261